fog/block_provider/src/mobilecoind.rs

```diff
@@ -0,0 +1,191 @@
+                            log::warn!(self.logger, "watcher is still behind on block index = {} after waiting {} seconds, caller will be blocked", block_index, watcher_timeout.as_secs());
+                            watcher_behind_timer = Instant::now();
+                        }
+                        sleep(POLL_BLOCK_TIMESTAMP_POLLING_FREQUENCY);
+                    }
+                    TimestampResultCode::BlockIndexOutOfBounds => {
+                        log::warn!(self.logger, "block index {} was out of bounds, we should not be scanning it, we will have junk timestamps for it", block_index);
+                        return u64::MAX;
+                    }
+                    TimestampResultCode::Unavailable => {
+                        log::crit!(self.logger, "watcher configuration is wrong and timestamps will not be available with this configuration. caller is blocked at block index {}", block_index);
+                        sleep(POLL_BLOCK_TIMESTAMP_ERROR_RETRY_FREQUENCY);
+                    }
+                    TimestampResultCode::WatcherDatabaseError => {
+                        log::crit!(self.logger, "The watcher database has an error which prevents us from getting timestamps. caller is blocked at block index {}", block_index);
+                        sleep(POLL_BLOCK_TIMESTAMP_ERROR_RETRY_FREQUENCY);
+                    }
+                    TimestampResultCode::TimestampFound => {
+                        return response.timestamp;
+                    }
+                    other => {
+                        log::crit!(
+                            self.logger,
+                            "Unexpected result code {:?} for block index {}",
+                            other,
+                            block_index
+                        );
+                        sleep(POLL_BLOCK_TIMESTAMP_ERROR_RETRY_FREQUENCY);
+                    }
+                },
+                Err(err) => {
+                    log::error!(
+                            self.logger,
+                            "Could not obtain timestamp for block {} due to error {}, this may mean the watcher is not correctly configured. will retry",
+                            block_index,
+                            err
+                        );
+                    sleep(POLL_BLOCK_TIMESTAMP_ERROR_RETRY_FREQUENCY);
+                }
+            };
+        }
+    }
+
+    fn get_tx_out_and_membership_proof_by_index(
+        &self,
+        tx_out_index: u64,
+    ) -> Result<(TxOut, TxOutMembershipProof), Error> {
+        let response = self
+            .client
+            .get_membership_proofs(&GetMembershipProofsRequest {
+                indices: vec![tx_out_index],
+                ..Default::default()
+            })?;
+
+        if response.output_list.len() != 1 {
+            log::error!(self.logger, "get_membership_proofs ")
+        }
+
+        let tx_out_with_proof = response.output_list.get(0).unwrap();
```

Comment:
Ha, nice catch although maybe not exactly what you had in mind. It looks like a return statement is missing in the `if response.output_list.len() != 1` block, so this would've totally just panicd. oops!